In [ ]:
from fastcore.all import *
from iomeval.readers import find_eval, load_evals
from mistocr.core import read_pgs

import httpx
import pandas as pd
from iomeval.pipeline import run_pipeline, batch_run, report_urls

In [ ]:
models = [
    'claude-sonnet-4-5',
    'claude-sonnet-4-6', 
    'claude-opus-4-5-20251101'
    ]

In [ ]:
evals_path = Path('../files/test/evaluations.json')
evals = load_evals(evals_path)
len(evals)

740

In [ ]:
evals[1]


### Meta-evaluation of IOM evaluations 2020–2024
**Year:** 2025 | **Organization:** IOM | **Countries:** Worldwide

**Documents:** 3 available  
**ID:** `2b297bc904ed1e542fbe67f62a8ea7cf`


## Find eval by url

In [ ]:
urls_elma = [
    {'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CBI%20Evaluation_final%20evaluation%20report.pdf',
    'title': 'Evaluation of IOM Cash-based Interventions'}]

In [ ]:
# Anne urls
urls_anne = [
    {
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf',
        'title':'EVALUATION OF THE IOM MIGRATION DATA STRATEGY 2020-2025'
    },
    {
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IAHE%20of%20the%20Drought%20Response%20in%20Somalia.pdf',
        'title': 'INTER-AGENCY HUMANITARIAN EVALUATION OF THE RESPONSE TO THE HUMANITARIAN CRISIS IN SOMALIA',
    },
    {
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOMFund%20evaluation_final%20report.pdf',
        'title': 'Fifth Evaluation of the IOM Development Fund'
    }
]

In [ ]:
for d in urls_elma:
    evl = find_eval(evals, d['url'], by='url')
    print(evl.id)

316c47c9dc19c820f3217434d4b93b28


In [ ]:
for d in urls_anne:
    evl = find_eval(evals, d['url'], by='url')
    print(evl.id)

9992310969aa2f428bc8aba29f865cf3


e21305a89aeb76fafa8f4270392d463a


0d4309b702e00faac631be8d609934ce


In [ ]:
ev_d = find_eval(evals, 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf', by='url')
ev_d


### EVALUATION OF IOM’S MIGRATION DATA STRATEGY
**Year:** 2025 | **Organization:** IOM | **Countries:** Worldwide

**Documents:** 6 available  
**ID:** `9992310969aa2f428bc8aba29f865cf3`


In [ ]:
def fix_report_url(evals, url, report_urls):
    ev = find_eval(evals, url, by='url')
    report_urls[ev.meta['Title']] = url
    return report_urls

In [ ]:
report_urls = fix_report_url(evals, 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf', report_urls)

In [ ]:
report_urls[ev_d.meta['Title']]

'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf'

## online reps

In [ ]:
rpts_online = [
    {
        'id': 0,
        'title': 'Better Migration Management Programme Phase III',
        'url':	'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Report%20without%20Annex_Evaluation%20IOM%20BMM%20Phase%20III_26-03-2025.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. BACKGROUND AND CONTEXT',
            '# 4. CONCLUSIONS',
            '# 5. RECOMMENDATIONS'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 1,
        'title': 'Evaluation of IOM Accountability to Affected Populations', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/AAP%20Evaluation%20Report_final_.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## Evaluation Context',
            '## Evaluation Purpose, Objectives and Scope',
            '# 3. CONCLUSIONS',
            '# 4. RECOMMENDATIONS'],
        'evaluator': 'Davide'
    },
    {
        'id': 2,
        'title': 'Final Evaluation of the EU-IOM Joint Initiative for migrant protection and reintegration in the Horn of Africa', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Final%20Evaluation%20Report%20Final_Olta%20NDOJA.pdf',
        'hoi': [
            '## Executive summary',
            '## Methodology',
            '## Findings',
            '## Conclusions and recommendations',
            '## 9. Conclusions',
            '## 10. Recommendations'],
        'evaluator': 'Davide'
    },
    {
        'id': 3,
        'title': 'Evaluation of Mental Health and Psychosocial Support in IOM', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Final%20Report%20Evaluation%20of%20MHPSS%20in%20IOM%208%202024.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## Objective of the Evaluation',
            '### ***Overall Objective of the Evaluation:***',
            '## Scope of the Evaluation',
            '# 4. CONCLUSIONS',
            '# 5. RECOMMENDATIONS'],
        'evaluator': 'Davide'
    },
    {
        'id': 4,
        'title': "Impact evaluation of the UN Secretary General's Peacebuilding Fund-supported East Darfur Assalaya-Sheiria-Yassin Triangle of Peace and Coexistence project", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IE139-PBF-PeaceFIELD-Darfur-report.pdf',
        'hoi': [
            '## Summary',
            '# 1. Introduction',
            '# 7. Discussion',
            '# 8. Conclusion'],
        'evaluator': 'Davide'
    },
    {
        'id': 5,
        'title': "EVALUATION OF IOM'S MIGRATION DATA STRATEGY", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf',
        'hoi': [],
        'evaluator': 'Davide'
    },
    {
        'id': 6,
        'title': 'MALDIVES: STRENGTHENING GOVERNMENT CAPACITY IN MIGRATION HEALTH POLICY DEVELOPMENT', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/EvaluationDocuments_Evaluation_Report_MA.0457_Maldives.pdf',
        'hoi': [
            '## SUMMARY OF THE EVALUATION ASSIGNMENT',
            '## 1. PURPOSE, SCOPE AND OBJECTIVES OF THE SERVICE',
            '## 2. APPROACH AND METHODOLOGY'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 7,
        'title': 'EX-POST EVALUATION OF THE PROJECT: SRI LANKA: UNDERSTANDING MIGRATION, ENVIRONMENTAL DEGRADATION AND CLIMATE CHANGE', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/EvaluationDocuments_NC.0046%20Evaluation%20Report_final.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## 1.1. CONTEXT AND PURPOSE OF THE EVALUATION',
            '## 1.2 EVALUATION BACKGROUND, SCOPE AND PURPOSE',
            '# 3. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 8,
        'title': 'Evaluation of the project “Regional Migration Policy and Knowledge Management Hub in South America”', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Evaluaci%C3%B3n%20PO.0175%20-%20Informe%20final.pdf',
        'hoi': [
            '# 1. EXECUTIVE SUMMARY',
            '## PROJECT BACKGROUND',
            '# 2. CONCLUSIONS',
            '# 3. RECOMMENDATIONS',
            '# 4. LESSONS LEARNED',
            '# 5. GOOD PRACTICES',
            '# 7. CONTEXT AND PURPOSE OF THE EVALUATION'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 9,
        'title': 'FINAL EVALUATION OF THE EU-IOM KNOWLEDGE MANAGEMENT HUB (KMH)', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/KMH%20Final%20Evaluation%20report_final_Olta%20NDOJA.pdf',
        'hoi': [
            '# Executive Summary',
            '# 1. Introduction',
            '# 2. The Evaluation',
            '# 4. Key lessons learned',
            '# 5. Conclusions',
            '# 6. Recommendations'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 10,
        'title': 'Final Internal Evaluation for Enhancing Knowledge on Remittances and Diaspora Engagement in South Sudan', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CD.0011%20evaluation%20report_final_Laura%20SMITH.pdf',
        'hoi': [
            '## Executive Summary',
            '## Evaluation Context and Purpose',
            '## Evaluation Methodology',
            '## Findings and Conclusions',
            '### Lessons learned:',
            '### Good practice:',
            '## Recommendations',
            '## Evaluation context and purpose',
            '## Evaluation Context',
            '### Evaluation Purpose',
            '## Evaluation Scope',
            '## Evaluation Criteria',
            '## Evaluation Framework and Methodology',
            '## Conclusions',
            '## Recommendations'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 11,
        'title': 'Midterm Evaluation of IOM East and Horn of Africa Regional Strategy 2020 – 2024', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Eval%20EHoA%20Regional%20Strategy%20Sept%202023%20final.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. CONTEXT AND PURPOSE OF THE EVALUATION',
            '# 3. LESSONS LEARNED AND GOOD PRACTICES',
            '# 4. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 12,
        'title': "THEMATIC EVALUATION OF IOM'S LABOUR MIGRATION AND MOBILITY STRATEGY AND INITIATIVES", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20LMI%20-%20final%20report%20April%2023.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. INTRODUCTION',
            '# 5. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Andres'
    }
]

### Pending updated repo API - in the meantime needs patching given not all in available list
def add_missing_urls():
    for o in rpts_online:
        report_urls.add(o['url'])

add_missing_urls()

In [ ]:
ids_online = [find_eval(evals, o['url'], by='url').id for o in rpts_online]
ids_online

['4341695461234eee3deb51ac68871109',
 '6c3c2cf3fa479112967612b0baddab72',
 '49d2fba781b6a7c0d94577479636ee6f',
 '22cac1c000836253adc445993e101560',
 'e8d9427f4a9094f68e17307faa567d8b',
 '9992310969aa2f428bc8aba29f865cf3',
 'a493c447cff3e03ca94024335d5c5c7a',
 '06fd0949dc874f388686461c19d2669c',
 '4c04ec0d3e651f7616190c6fc2dcfc18',
 '891448b7a4bf12763ea0579dd67e289d',
 '75bad4665ed458cce78740e3d86f72d0',
 'f409b3f9a0671ec69a8591c425a45c1d',
 '760568a15e144bcfb88dbbd50247dc51']

In [ ]:
base_path = Path('../../data')

In [ ]:
current = {o.stem for o in (base_path / 'results').ls()}
len(current)

133

In [ ]:
# Check nont of the online one are already processed
#set(ids_online) & current

In [ ]:
evs_2022 = [e.id for e in evals if e.meta['Year'] == 2022]
len(evs_2022)

67

In [ ]:
ids = set(current) & set(evs_2022)
len(ids)

35

In [ ]:
ids_todo = list(set(evs_2022) - current)
ids_todo

['ff24895e93304edf09d6fa5e50499ff4',
 'f0b09b92ea8ad6dddd9623de68a8d278',
 '1e3746e92137c5f70ffbb32a4e09bb4d',
 '53bbce227cb118b9316f33ac9b5fe8c0',
 'd3eec5e8a9b0d8bad5d84576cb4c4c96',
 '3274df5c61f151a9237bcf064ade86fc',
 'db6b87320edb3055c6aca614992fdaa1',
 'c244e298e865790fe29831ecf933c08e',
 'b9241cd2086d3c665b0c0678f3d56f69',
 '71eb96b03c1ac851a81bf4875776be15',
 'a019206cb9f84bd0dd0e488f44111318',
 '2cae361c6779b561af07200e3d4e4051',
 '670b60f173c3a1c58a604436b6d2f50c',
 'cb179f92dfec7102c26fe6f2d55de237',
 'fa59908612a9a8ef306b310a8f00601e',
 'f7b2bfd11cf04b407fcbe52c9606a8e6',
 'cad8e2d8d7e01f33f82b6a0956853b11',
 '6b39844a266cb0a01b85f6be15c14bdf',
 '93e51fcbedaaddd6a7e38fd8ee039614',
 '926fd1567e78ae9a661fc93ecf81418d',
 '04daa54157c3cd7221b89c2fd6466167',
 '67606e99aaa89670f447141ea92f5bf8',
 '407b8f004272c8d4276305976cb3316b',
 '2ea5b130f7a8024d1275b12ee428066a',
 '2ceadc314417082154abb0c3e004672d',
 '91c424f1e812a8e26f07d7896a327f6f',
 '72840a765c5a52032a1f863a0d038ba3',
 

In [ ]:
len(ids_todo)

32

In [ ]:
prompt_fix_hdgs_v2 = """You are cleaning markdown headings extracted from OCR'd evaluation reports. These reports follow a general structure (with variation):

TYPICAL STRUCTURE:
- Title (H1 — the ONLY H1)
- Executive Summary (H2)
- Table of Contents (H2) — may appear before or after Executive Summary
- List of Figures / List of Abbreviations (H2)
- Introduction / Background (H2)
- Findings (H2)
- Conclusions (H2)
- Recommendations (H2)
- Annexes (H2)

CRITICAL: DISTINGUISHING TOC ITEMS FROM REAL HEADINGS
Table of Contents entries list the document's sections but are NOT actual section headings. Clues that something is a TOC item (not a real heading):
- It appears between "Table of Contents" and the next major section (e.g., "Introduction")
- Multiple items appear in sequence on the same or consecutive pages with page number references
- The TOC heading text will be REPEATED later as an actual heading
→ TOC items should be converted to regular text: output the text WITHOUT any # symbols

MISCLASSIFIED HEADINGS (output as "" to remove, or without # to convert to text):
- Table titles (e.g., "Table 1: Budget Overview")
- Figure captions (e.g., "Figure 3.1: Migration flows")
- Bold text used for emphasis, not structure
- Items in bulleted/numbered lists misidentified as headings
- Page headers/footers repeated across pages

HEADING HIERARCHY RULES:
1. **Single H1**: Only the document title should be #. Everything else is ## or deeper.

2. **Numbered sections imply hierarchy**:
   - "1. Introduction" (##) → "1.1 Background" (###) → "1.1.1 Context" (####)
   - Parent sections MUST be shallower than children
   - Works with any style: 1/1.1, A/A.1, I/I.A, etc.
   
   *Example error:*
   ```
   ## 3. Section title
   ## 3.1 Subsection title  ← WRONG: should be ### because 3.1 is a child of 3
   ```
   
   The number of segments determines depth:
   - "3" (1 segment) → parent level
   - "3.1" (2 segments) → one level deeper than parent
   - "3.1.1" (3 segments) → two levels deeper than parent

3. **No level jumps**: Going deeper can only increase by one # at a time
   - Wrong: ## → ##### 
   - Right: ## → ###

4. **Unnumbered headings in numbered documents**: If the document uses consistent numbering, an unnumbered "heading" within that structure is likely misclassified bold text → remove the # symbols

5. **Mixed or no numbering**: Use semantic judgment based on content and typical report structure above

ANALYSIS APPROACH:
1. First, identify the title (should be the only H1)
2. Locate the TOC section — everything between "Table of Contents"/"Contents" heading and the next major section is likely TOC items, not headings
3. For each heading, determine: Is this a real structural heading, a TOC item, or misclassified text?
4. For real headings, verify the # depth matches the implied hierarchy from numbering or semantic role
"""


In [ ]:
#id_failed = ["f254cc8d8ed169ab7b7fd11fad6d936e"]
id_single = ['9992310969aa2f428bc8aba29f865cf3']

In [ ]:
models[0]

'claude-sonnet-4-5'

In [ ]:
result = await batch_run(
    evals, report_urls, 
    base_path='../../data', 
    #ids=list(ids_todo)[:10],
    ids=id_single,
    add_img_desc=False,
    fix_kwargs=dict(
        model=models[0], 
        prompt=prompt_fix_hdgs_v2,
        reasoning_effort='low'  # or 'low', 'medium'
        ),
    verbose=False,
    model=models[0]
)

iomeval.pipeline - INFO - 99923109... awaiting_curation


In [ ]:
ids = [
    '316c47c9dc19c820f3217434d4b93b28',
    '9992310969aa2f428bc8aba29f865cf3',
    'e21305a89aeb76fafa8f4270392d463a',
    '0d4309b702e00faac631be8d609934ce']

In [ ]:
!ls ../../data/results

002673527ea7ecd54b8f1b99c42a66a8.json


00a7bb8b91704acad9a7127016a14404.json


00adef76065e408395a65c9a162037e1.json


03a49c3a600b3f4af16b9067643ebd30.json


05e79e26c55262a761b22c0689ea42cd.json


06fd0949dc874f388686461c19d2669c.json


0864fde6ee108ae4fd86dc4ec961a7ba.json


091f52fee51e6b07bf11fa49fcfffa15.json


0d4309b702e00faac631be8d609934ce.json


0d4309b702e00faac631be8d609934ce.json.bak


0e812bc6946e1dada75db935ff4f57b7.json


149308aa166dd79bfa3ef04b6c7b1f74.json


155ced3cb214d66c6c1803ccc213c5bc.json


162cb7492199e3be5ae17afa9c58d2be.json


16b8975171a68a257025887a81940b6a.json


1700c8dbadc3d87d6911a8ccc6d18b63.json


1724ba26d5f0361462ab2310889693ea.json


1896e29ade1e9a98b43316a446acb295.json


18d02982f0a3b346ddd612bdc8993693.json


1a57974ab89d7280988aa6b706147ce1.json


1d57797a87cfcc419146ad546945fa92.json


213921c4d449fac63539e192d7ffa5c3.json


227dbf5b243f3973312b7de3be53105d.json


22cac1c000836253adc445993e101560.json


231001cdfad75231a6ee6a68dadbe2ac.json


2af8683730af0c8fe704a48b3040527d.json


2b297bc904ed1e542fbe67f62a8ea7cf.json


2e376c288e2428bb2fc7c6517a15b76e.json


2fc235bc161ea39a68696322dd58552b.json


2ffecb818f3b24bda69c24da68524bbf.json


316c47c9dc19c820f3217434d4b93b28.json


316c47c9dc19c820f3217434d4b93b28.json.bak


34b11afa846de3bd7df009fa30a4d847.json


382763d830da81737caaa9aed9a17b1a.json


3d2c35aec5bde93fc1db65884af3daa8.json


3eac0012a1e2d458a8a6ff597d19e988.json


431045743ac88ac6fe2770cfaa41abc6.json


4341695461234eee3deb51ac68871109.json


4706bc3d429dac3debe7d52edcba620a.json


48fb10f3895036743cb8811daac46bcd.json


49d2fba781b6a7c0d94577479636ee6f.json


4c04ec0d3e651f7616190c6fc2dcfc18.json


4f578e46c1e5d12f86172bba1b96eb64.json


50ae07f4997dfefcf7cb5f27b9a681b5.json


520b24ab88bebd67a534796fd3b6ceca.json


5416c48dc6d986f08d438cb15d697d7a.json


5641b454ff68b94119938cd4e3014024.json


56e6f2bc9f55a0c6d07341ddde1283db.json


56edcf23135e9a78a854b252f6b389b3.json


57ef0f5e04ec1eaa6c45a625e7d4d541.json


58197a22fa56c40128dbda1ffca25c61.json


5e42535311296a981c171e9411e29b5d.json


5eeab35646099364f9f676bd79150cd5.json


645fd84a07744de4acc076016768d654.json


67198e6af9cb7368a68ce9ca5d77ddb9.json


68a24be2dce04ce1ff31f03e1f7a3eb9.json


68fb4700d5cc22b655d202ba310dac60.json


6b30198a6f8b83c8cbf36ef7d9e8cb29.json


6c3c2cf3fa479112967612b0baddab72.json


7363e7dce1569640d3093e30d02919b2.json


75bad4665ed458cce78740e3d86f72d0.json


760568a15e144bcfb88dbbd50247dc51.json


7a6f58a2fa72c3290ceae54f0924ea18.json


7ac1b169fd2c594e5f3411f1fca1fdb1.json


7c89660d359efcf0f8bba26f79ae8071.json


7d0d67d0123b23c0683214c0846bcfb4.json


7eb413021009f18b1d7f62c896cdf841.json


8044d6df2b70eb49de18978a8ed30045.json


804cbff9b801da925aca9e7cd1827195.json


84be65a7b7c671a3ff0ada8678ed69c7.json


855ca10090c95f1ba17555c7ca2ace65.json


8579dbefab619193e28e4975d7c8a884.json


86784f69c04474def48029644d5039e1.json


891448b7a4bf12763ea0579dd67e289d.json


8b846b0820a1122f7984b8bfecdb233d.json


8cf2c9b020b1d01d27cfd2bd8c86a29a.json


8dd2261100ff5d714e73b411eed043d5.json


8e6d1ed85f67b6ef6f27d7826cd28eba.json


8f27e97343fa61f470180e15fc12aa11.json


8faad6edb0414cfc8f35415a93726e36.json


9051a78159bcf16ca58130fc8b7b81d3.json


935a5d3af609e72dc8db68968679d66e.json


94852dfaa9e277db14372f5b85a5412a.json


9558482be1b049827f630c1760025c0f.json


9591f983c436297929b9e240a3c28df6.json


96d833bc0d1b2f20e3a5cafcc0a3fef7.json


99088dbe6e81a38af6dab8499fa57463.json


9992310969aa2f428bc8aba29f865cf3.json


9992310969aa2f428bc8aba29f865cf3.json.bak


99b614cfeb34acd1eae8964ae3c93eb0.json


9bfd29a6c84de3f84537c7542c251dc6.json


9fec9ba7a298c9dd11ae2204403409d2.json


a0ac79f7126465063a6678a434d45903.json


a1df1fcfe0e056a896dd0671d7a675d2.json


a25b7c619be8945fcbf1da8911c298c6.json


a3702206d8688b26655616b1b47e60da.json


a493c447cff3e03ca94024335d5c5c7a.json


a4e7b40c5818539493cbf705ebc082ed.json


a8507cd92c7c2101e2fac95f52a5fb0d.json


a9dea21fd254df7759b3936903e0a885.json


af04aae904bafb85949fc3c9d84c4f90.json


b00abee5849be373ec567ef5e09032fb.json


b1089282146717912f61e3f7a9c05e19.json


b4aade6b2a658923c1e15fdf22aee44f.json


be58de37a0f5d78299f542ecaadd08a8.json


c352e1bdfac83b00b53d7fe420b2135c.json


c55e1e07ac5df5c8060a084cd811de6e.json


c5b09a01c1769c0d50bfe4116b98afa8.json


c660e774d14854e20dc74457712b50ec.json


c7206c0b501aa34d456617ca618c5aff.json


ca72ad8607870852b14c9cc76961a4d4.json


cf4fcf44e2eff305339aeeb5d2290f7f.json


cfd82db88226ac1a8e6ccb710c3f5c6d.json


d1297e509b8befc99eca8f70f4abad0c.json


d4ea4639d798bb9677b5ca07a51d0d04.json


d5476e635b658379f25e05ecd77d8095.json


d65ed11dd200fca6216af40f18d8c190.json


d68a6f7f510ab6104bb2fcde51de1dc2.json


d6cc6a5aadb321aff0ff993a15b0f122.json


d9a4f862e120eea65466ce0612254b12.json


db2a446ee2bf1566c1264769e4aa9dff.json


dba9eb1220368d28682d8531108e767d.json


dfe733991ff0a81b4c4bcac609c350f3.json


e208c67542e6ff9e02f49599bb0b127b.json


e21305a89aeb76fafa8f4270392d463a.json


e21305a89aeb76fafa8f4270392d463a.json.bak


e4cc1f9d938aa24966e7f1e8de790a05.json


e5287e244baa3e1fa7f2a6e72f9d1fc4.json


e5a55f0235e0a312dcf2f09ba3c385a1.json


e8d9427f4a9094f68e17307faa567d8b.json


e9e1c75dfab81073d2d7f0a04190d344.json


eb2a3511e23f66713c6c53d678f6a659.json


f254cc8d8ed169ab7b7fd11fad6d936e.json


f409b3f9a0671ec69a8591c425a45c1d.json


f77aa21a5f4e5cdb9cc1a863a2271800.json


f9ff95796dd18443fcc94fd6b795acf6.json


fc71c41bead65af73ba2f86db3a808e8.json


ffa7846d8ad478ed9fa22411e39fc685.json


These $`ids` have already been downloaded, ocrd, curated, ... and now ready for mapping how to do so? Can you check the "07_pipeline.ipynb"  to inform your response? You also get access to the result json for each report to check their status

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠

Let me look at the pipeline notebook and check the result files for those IDs.

<details class='tool-usage-details'>
<summary>find_msgs(dname='/iomeval/nbs/07_pipeline', headers_only=True)</summary>

```json
{
  "id": "toolu_01BNLpy8QqM91fmmBWnFEdm4",
  "call": {
    "function": "find_msgs",
    "arguments": {
      "dname": "/iomeval/nbs/07_pipeline",
      "headers_only": "True"
    }
  },
  "result": "<TRUNCATED>\u2026c1351103', 'time_run': '', 'is_exported': 0, 'skipped': 0, 'hidden': False, 'bookmark': {}, 'i_collapsed': 0, 'o_collapsed': 0, 'heading_collapsed': False, 'i_clamp': False, 'o_clamp': False, 'pinned': 0, 'run': {}, 'content': '# Pipeline\\n<TRUNCATED>', 'output': '', 'msg_type': 'note', 'input_tokens': 38, 'output_tokens': {}}, {'id': '_de639b56', 'time_run': '', 'is_exported': 0, 'skipped': 0, 'hidden': False, 'bookmark': {}, 'i_collapsed': 0, 'o_collapsed': 0, 'heading_collapsed': False, 'i_clamp': False, 'o_clamp': False, 'pinned': 0, 'run': {}, 'content': '## Report class\\n<TRUNCATED>', 'output': '', 'msg_type': 'note', 'input_tokens': 54, 'output_tokens': {}}, {'id': '_d896e3d2', 'time_run': '', 'is_exported': 0, 'skipped': 0, 'hidden': False, 'bookmark': {}, 'i_collapsed': 0, 'o_collapsed': 0, 'heading_collapsed': False, 'i_clamp': False, 'o_clamp': False, 'pinned': 0, 'run': {}, 'content': '#### Creating reports\\n<TRUNCATED>', 'output': '', 'msg_type': 'note'\u2026</TRUNCATED>"
}
```

</details>



<details class='tool-usage-details'>
<summary>python(code='<…>…e status of each result file\ni…</…>')</summary>

```json
{
  "id": "toolu_01F5agDUR1tFzi6Ab8u7PD1L",
  "call": {
    "function": "python",
    "arguments": {
      "code": "# Check the status of each result file\nimport json\n\nresults = {}\nfor eid in ['316c47c9dc19c820f3217434d4b93b28', '9992310969aa2f428bc8aba29f865cf3', 'e21305a89aeb76fafa8f4270392d463a', '0d4309b702e00faac631be8d609934ce']:\n    p = Path(f'../../data/results/{eid}.json')\n    if p.exists():\n        data = json.loads(p.read_text())\n        results[eid] = {k: v for k, v in data.items() if k in ('status', 'title', 'steps_completed', 'current_step')}\n    else:\n        results[eid] = 'NOT FOUND'\n\nfor eid, info in results.items():\n    print(f\"\\n{eid}:\")\n    if isinstance(info, dict):\n        for k,v in info.items(): print(f\"  {k}: {v}\")\n    else:\n        print(f\"  {info}\")"
    }
  },
  "result": "{'warnings': '`_write_` is not available in this sandbox; ask the user to add it to the available tools'}"
}
```

</details>



🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠

Let me check the result files and the relevant pipeline sections.

<details class='tool-usage-details'>
<summary>view(path='<…>…/results/316c47c9dc19c820f3217…</…>', view_range=[1, 50])</summary>

```json
{
  "id": "toolu_01NWLVwcuYN7EgKbZpHXTYSU",
  "call": {
    "function": "view",
    "arguments": {
      "path": "../../data/results/316c47c9dc19c820f3217434d4b93b28.json",
      "view_range": "[1, 50]"
    }
  },
  "result": "<TRUNCATED>\u2026: \"316c47c9dc19c820f3217434d4b93b28\",\\n  \"report_url\": \"https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CBI%20Evaluation_final%20evaluation%20report.pdf\",\\n  \"meta\": {\\n    \"Title\": \"EVALUATION OF IOM CASH-BASED INTERVENTIONS (CBI)\",\\n    \"Year\": 2025,\\n    \"Author\": \"IOM CENTRAL EVALUATION\",\\n    \"Best Practicesor Lessons Learnt\": \"Yes\",\\n    \"Date of Publication\": \"2025-03-28\",\\n    \"Donor\": \"IOM\",\\n    \"Evaluation Brief\": \"Yes\",\\n    \"Evaluation Commissioner\": \"IOM\",\\n    \"Evaluation Coverage\": \"Global\",\\n    \"Evaluation Period From Date\": NaN,\\n    \"Evaluation Period To Date\": NaN,\\n    \"Executive Summary\": \"Yes\",\\n    \"External Version of the Report\": \"No\",\\n    \"Languages\": \"English\",\\n    \"Migration Thematic Areas\": \"Migration and Development - remittances and financial empowerment, Multiple thematic overview, Organisational policy/strategy\",\\n    \"Name of Project(s) Being Evaluated\": NaN,\\n    \"Number of Pages Excluding annexes\": NaN,\u2026</TRUNCATED>"
}
```

</details>



<details class='tool-usage-details'>
<summary>find_msgs(dname='/iomeval/nbs/07_pipeline', header_section='## Thematic mapping')</summary>

```json
{
  "id": "toolu_01NYzJuNmepX89DVRqcDvDss",
  "call": {
    "function": "find_msgs",
    "arguments": {
      "dname": "/iomeval/nbs/07_pipeline",
      "header_section": "## Thematic mapping"
    }
  },
  "result": "<TRUNCATED>\u2026e0506f62', 'time_run': '', 'is_exported': 0, 'skipped': 0, 'hidden': False, 'bookmark': {}, 'i_collapsed': 0, 'o_collapsed': 0, 'heading_collapsed': False, 'i_clamp': False, 'o_clamp': False, 'pinned': 0, 'run': {}, 'content': \"## Thematic mapping\\n\\nMap extracted sections to IOM's strategic frameworks (SRF and GCM). Each mapping method can be run independently after curation.\", 'output': '', 'msg_type': 'note', 'input_tokens': 56, 'output_tokens': {}}, {'id': '_f7b5bf6c', 'time_run': '', 'is_exported': 1, 'skipped': 0, 'hidden': False, 'bookmark': {}, 'i_collapsed': 0, 'o_collapsed': 0, 'heading_collapsed': False, 'i_clamp': False, 'o_clamp': False, 'pinned': 0, 'run': {}, 'content': '@patch\\ndef ensure_sys_blocks(self:Report) -> None:\\n    \"Ensure system blocks are available\"\\n    if not hasattr(self, \\'_sys_blocks\\'): self._sys_blocks = mk_system_blocks(self.get_sections())', 'output': '', 'msg_type': 'code', 'input_tokens': 70, 'output_tokens': {}}, {'id': '_5f5\u2026</TRUNCATED>"
}
```

</details>



🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠

Let me also check the status in those result files and look at the full pipeline run section.

<details class='tool-usage-details'>
<summary>find_msgs(dname='/iomeval/nbs/07_pipeline', header_section='## Run full pipeline')</summary>

```json
{
  "id": "toolu_0121EYQEd98XpTVPLwGKbT1r",
  "call": {
    "function": "find_msgs",
    "arguments": {
      "dname": "/iomeval/nbs/07_pipeline",
      "header_section": "## Run full pipeline"
    }
  },
  "result": "<TRUNCATED>\u202625ad7c8a', 'time_run': '', 'is_exported': 0, 'skipped': 0, 'hidden': False, 'bookmark': {}, 'i_collapsed': 0, 'o_collapsed': 0, 'heading_collapsed': False, 'i_clamp': False, 'o_clamp': False, 'pinned': 0, 'run': {}, 'content': '## Run full pipeline\\n\\nRun the pipeline on a single evaluation report. The pipeline runs until curation is needed, then continues after curation is complete. Re-run the same call after using the curator app to complete all stages.', 'output': '', 'msg_type': 'note', 'input_tokens': 71, 'output_tokens': {}}, {'id': '_b63be3f5', 'time_run': '', 'is_exported': 1, 'skipped': 0, 'hidden': False, 'bookmark': {}, 'i_collapsed': 0, 'o_collapsed': 0, 'heading_collapsed': False, 'i_clamp': False, 'o_clamp': False, 'pinned': 0, 'run': {}, 'content': 'def should_force(force,     # Bool to force all steps, or set of step names to force\\n                 step       # Step name to check\\n                ) -> bool:  # Whether to force the step\\n    \"Check i\u2026</TRUNCATED>"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…s" ../../data/results/316c47c9…</…>')</summary>

```json
{
  "id": "toolu_01MfyV1xcG5HfQ5XYBvRXLNS",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-m1 \"status\" ../../data/results/316c47c9dc19c820f3217434d4b93b28.json"
    }
  },
  "result": "'64:  \"curation_status\": \"sections_selected\",\\n'"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…s" ../../data/results/99923109…</…>')</summary>

```json
{
  "id": "toolu_01JHo9ecyM9fqAMYtUHNNG93",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-m1 \"status\" ../../data/results/9992310969aa2f428bc8aba29f865cf3.json"
    }
  },
  "result": "'74:  \"curation_status\": \"sections_selected\",\\n'"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…s" ../../data/results/e21305a8…</…>')</summary>

```json
{
  "id": "toolu_018zyVry2D7zzLGgGr4xKP7y",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-m1 \"status\" ../../data/results/e21305a89aeb76fafa8f4270392d463a.json"
    }
  },
  "result": "'54:  \"curation_status\": \"sections_selected\",\\n'"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…s" ../../data/results/0d4309b7…</…>')</summary>

```json
{
  "id": "toolu_01TZ6L4cxfqvu6S96VNHXcCf",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-m1 \"status\" ../../data/results/0d4309b702e00faac631be8d609934ce.json"
    }
  },
  "result": "'64:  \"curation_status\": \"sections_selected\",\\n'"
}
```

</details>



🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠🧠

Let me check if these reports already have mappings completed.

<details class='tool-usage-details'>
<summary>rg(argstr='<…>…id" ../../data/results/316c47c…</…>')</summary>

```json
{
  "id": "toolu_01QZs8ReFmLbUSDxoWrC7qdv",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-c \"theme_id\" ../../data/results/316c47c9dc19c820f3217434d4b93b28.json"
    }
  },
  "result": "''"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…id" ../../data/results/9992310…</…>')</summary>

```json
{
  "id": "toolu_01Dirax6sRywpa6pk5fGqVSy",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-c \"theme_id\" ../../data/results/9992310969aa2f428bc8aba29f865cf3.json"
    }
  },
  "result": "''"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…id" ../../data/results/e21305a…</…>')</summary>

```json
{
  "id": "toolu_01ETEd3wMU5NeBbHKP66Ey3J",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-c \"theme_id\" ../../data/results/e21305a89aeb76fafa8f4270392d463a.json"
    }
  },
  "result": "''"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…id" ../../data/results/0d4309b…</…>')</summary>

```json
{
  "id": "toolu_0133BcKKpnczsBNzoToxH9mM",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-c \"theme_id\" ../../data/results/0d4309b702e00faac631be8d609934ce.json"
    }
  },
  "result": "''"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…ngs" ../../data/results/316c47…</…>')</summary>

```json
{
  "id": "toolu_01BW4SFRmyNK1ULoTShdVoeK",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-m1 \"mappings\" ../../data/results/316c47c9dc19c820f3217434d4b93b28.json"
    }
  },
  "result": "'71:  \"mappings\": {},\\n'"
}
```

</details>



<details class='tool-usage-details'>
<summary>rg(argstr='<…>…ngs" ../../data/results/999231…</…>')</summary>

```json
{
  "id": "toolu_013x2Kv5Vs6YtUYFHaUHAY6N",
  "call": {
    "function": "rg",
    "arguments": {
      "argstr": "-m1 \"mappings\" ../../data/results/9992310969aa2f428bc8aba29f865cf3.json"
    }
  },
  "result": "'83:  \"mappings\": {},\\n'"
}
```

</details>

Now I have a clear picture. Here's the status of your 4 reports:

| ID (short) | Curation | Mappings |
|---|---|---|
| `316c47c9...` (CBI) | ✅ `sections_selected` | ❌ empty `{}` |
| `9992310...` (Migration Data Strategy) | ✅ `sections_selected` | ❌ empty `{}` |
| `e213058...` (Somalia Drought IAHE) | ✅ `sections_selected` | ❌ empty `{}` |
| `0d4309b...` (IOM Dev Fund) | ✅ `sections_selected` | ❌ empty `{}` |

All 4 are **curated** (headings selected) but have **no thematic mappings** yet. The next step is to run the mapping stage. Based on the pipeline notebook, since curation is done, `run_pipeline` will skip straight to the mapping calls (`map_enbs` → `map_ccps` → `map_gcms` → `map_outs`).

You can run mappings on all 4 with `batch_run` using those IDs, similar to what you did earlier for the single report:

```python
result = await batch_run(
    evals, report_urls, 
    base_path='../../data', 
    ids=ids,
    add_img_desc=False,
    fix_kwargs=dict(
        model=models[0], 
        prompt=prompt_fix_hdgs_v2,
        reasoning_effort='low'
    ),
    verbose=True,
    model=models[0]
)
```

This should pick up from where curation left off and run all 4 mapping stages for each report. Want to go ahead with this?

In [ ]:
from shutil import copy2
for eid in ids:
    src = base_path / 'results' / f'{eid}.json'
    copy2(src, src.with_suffix('.json.bak'))

In [ ]:
result = await batch_run(
    evals, report_urls, 
    base_path='../../data', 
    ids=ids,
    add_img_desc=False,
    fix_kwargs=dict(
        model=models[0], 
        prompt=prompt_fix_hdgs_v2,
        reasoning_effort='low'
    ),
    verbose=True,
    model=models[0]
)

iomeval.pipeline - INFO - Mapping enablers...


iomeval.pipeline - INFO - Mapping CCPs...


iomeval.pipeline - INFO - Mapping GCM objectives...


iomeval.pipeline - INFO - Mapping outputs...


iomeval.pipeline - INFO - Pipeline complete!


iomeval.pipeline - INFO - 99923109... completed


iomeval.pipeline - INFO - Mapping enablers...


iomeval.pipeline - INFO - Mapping CCPs...


iomeval.pipeline - INFO - Mapping GCM objectives...


iomeval.pipeline - INFO - Mapping outputs...


iomeval.pipeline - INFO - Pipeline complete!


iomeval.pipeline - INFO - 316c47c9... completed


iomeval.pipeline - INFO - Mapping enablers...


iomeval.pipeline - INFO - Mapping CCPs...


iomeval.pipeline - INFO - Mapping GCM objectives...


iomeval.pipeline - INFO - Mapping outputs...


iomeval.pipeline - INFO - Pipeline complete!


iomeval.pipeline - INFO - e21305a8... completed


iomeval.pipeline - INFO - Mapping enablers...


iomeval.pipeline - INFO - Mapping CCPs...


iomeval.pipeline - INFO - Mapping GCM objectives...


iomeval.pipeline - INFO - Mapping outputs...


iomeval.pipeline - INFO - Pipeline complete!


iomeval.pipeline - INFO - 0d4309b7... completed


In [ ]:
!ls ../../data

batch_runs  md	pdf  results  results-bu  results_new  results_test


In [ ]:
from shutil import copy2
for eid in ids:
    src = base_path / 'results' / f'{eid}.json'
    dst = base_path / 'results_new' / f'{eid}.json'
    copy2(src, dst)
    print(f"Copied {eid}")

Copied 316c47c9dc19c820f3217434d4b93b28


Copied 9992310969aa2f428bc8aba29f865cf3


Copied e21305a89aeb76fafa8f4270392d463a


Copied 0d4309b702e00faac631be8d609934ce


## Monitor batch runs

In [ ]:
import json

def load_all_batch_runs(base_path):
    """Load all batch run JSON files from the batch_runs folder."""
    return {p.stem: json.loads(p.read_text()) for p in (base_path / 'batch_runs').ls()}

def get_all_failed(base_path):
    """Return dict of {batch_name: [failed_ids]} for all runs with failures."""
    return {name: br['failed'] for name, br in load_all_batch_runs(base_path).items() if br.get('failed')}

def find_status(base_path, id):
    """Return list of (batch_name, status, error) tuples for an ID across all batch runs."""
    results = []
    for name, br in load_all_batch_runs(base_path).items():
        for status in ['completed', 'awaiting_curation', 'skipped']:
            if id in br.get(status, []):
                results.append((name, status, None))
        for failed in br.get('failed', []):
            if failed.get('id') == id:
                results.append((name, 'failed', failed.get('error')))
    return results

In [ ]:
failed = get_all_failed(base_path)
len(failed), failed

(6,
 {'2026-02-12_10-32': [{'id': '149308aa166dd79bfa3ef04b6c7b1f74',
    'step': 'ocr',
    'error': "mistocr.refine.fix_hdgs() got multiple values for keyword argument 'model'"},
   {'id': '57ef0f5e04ec1eaa6c45a625e7d4d541',
    'step': 'ocr',
    'error': "mistocr.refine.fix_hdgs() got multiple values for keyword argument 'model'"}],
  '2026-02-11_13-49': [{'id': 'cf4fcf44e2eff305339aeeb5d2290f7f',
    'step': 'ocr',
    'error': 'API error occurred: Status 413 Content-Type "application/json; charset=utf-8". Body: {\n  "message":"Request size limit exceeded",\n  "request_id":"512b3f1547d3a17be0d009134810a156"\n}'}],
  '2026-02-13_09-17': [{'id': '00a7bb8b91704acad9a7127016a14404',
    'step': 'ocr',
    'error': 'API error occurred: Status 401 Content-Type "application/json; charset=utf-8". Body: {"detail":"Unauthorized"}'}],
  '2026-02-12_10-33': [{'id': 'f254cc8d8ed169ab7b7fd11fad6d936e',
    'step': 'ocr',
    'error': "mistocr.refine.fix_hdgs() got multiple values for keyword ar

In [ ]:
find_status(base_path, "8f27e97343fa61f470180e15fc12aa11")

[('2026-02-13_09-13',
  'failed',
  'API error occurred: Status 401 Content-Type "application/json; charset=utf-8". Body: {"detail":"Unauthorized"}')]

In [ ]:
import json

# Get all unique failed IDs
failed_ids = {f['id'] for fs in failed.values() for f in fs}

# Check which ones don't have curation_status == "sections_selected"
not_processed = []
for fid in failed_ids:
    p = base_path / 'results' / f'{fid}.json'
    if not p.exists():
        not_processed.append((fid, 'no_results_file'))
    else:
        data = json.loads(p.read_text())
        status = data.get('curation_status', 'pending')
        if status != 'sections_selected':
            not_processed.append((fid, status))

not_processed

[('68a24be2dce04ce1ff31f03e1f7a3eb9', 'pending'),
 ('091f52fee51e6b07bf11fa49fcfffa15', 'pending'),
 ('e9e1c75dfab81073d2d7f0a04190d344', 'pending'),
 ('8f27e97343fa61f470180e15fc12aa11', 'pending'),
 ('149308aa166dd79bfa3ef04b6c7b1f74', 'pending'),
 ('00a7bb8b91704acad9a7127016a14404', 'pending'),
 ('dba9eb1220368d28682d8531108e767d', 'pending'),
 ('16b8975171a68a257025887a81940b6a', 'pending'),
 ('57ef0f5e04ec1eaa6c45a625e7d4d541', 'pending'),
 ('8579dbefab619193e28e4975d7c8a884', 'pending'),
 ('155ced3cb214d66c6c1803ccc213c5bc', 'pending'),
 ('cf4fcf44e2eff305339aeeb5d2290f7f', 'pending'),
 ('03a49c3a600b3f4af16b9067643ebd30', 'pending'),
 ('fc71c41bead65af73ba2f86db3a808e8', 'pending'),
 ('84be65a7b7c671a3ff0ada8678ed69c7', 'pending')]

In [ ]:
ids_to_retry = [fid for fid, _ in not_processed]
ids_to_retry

['68a24be2dce04ce1ff31f03e1f7a3eb9',
 '091f52fee51e6b07bf11fa49fcfffa15',
 'e9e1c75dfab81073d2d7f0a04190d344',
 '8f27e97343fa61f470180e15fc12aa11',
 '149308aa166dd79bfa3ef04b6c7b1f74',
 '00a7bb8b91704acad9a7127016a14404',
 'dba9eb1220368d28682d8531108e767d',
 '16b8975171a68a257025887a81940b6a',
 '57ef0f5e04ec1eaa6c45a625e7d4d541',
 '8579dbefab619193e28e4975d7c8a884',
 '155ced3cb214d66c6c1803ccc213c5bc',
 'cf4fcf44e2eff305339aeeb5d2290f7f',
 '03a49c3a600b3f4af16b9067643ebd30',
 'fc71c41bead65af73ba2f86db3a808e8',
 '84be65a7b7c671a3ff0ada8678ed69c7']

In [ ]:
result = await batch_run(
    evals, report_urls, 
    base_path='../../data', 
    ids=ids_to_retry,
    force={'ocr'},  # Force just the OCR step
    add_img_desc=False,
    fix_kwargs=dict(
        model='claude-opus-4-5', 
        prompt=prompt_fix_hdgs_v2,
        reasoning_effort='low'
    ),
    verbose=False,
    model='claude-sonnet-4-5'
)

iomeval.pipeline - INFO - cf4fcf44... failed


iomeval.pipeline - INFO - 149308aa... awaiting_curation


iomeval.pipeline - INFO - 57ef0f5e... awaiting_curation


iomeval.pipeline - INFO - 8579dbef... awaiting_curation


iomeval.pipeline - INFO - dba9eb12... awaiting_curation


iomeval.pipeline - INFO - e9e1c75d... awaiting_curation


iomeval.pipeline - INFO - fc71c41b... awaiting_curation


iomeval.pipeline - INFO - 00a7bb8b... awaiting_curation


iomeval.pipeline - INFO - 84be65a7... awaiting_curation


iomeval.pipeline - INFO - 155ced3c... awaiting_curation


iomeval.pipeline - INFO - 68a24be2... awaiting_curation


iomeval.pipeline - INFO - 03a49c3a... awaiting_curation


iomeval.pipeline - INFO - 16b89751... awaiting_curation


iomeval.pipeline - INFO - 091f52fe... awaiting_curation


iomeval.pipeline - INFO - 8f27e973... awaiting_curation


## Legacy

In [ ]:
#result = await batch_run(evals, report_urls, base_path='../../data', ids=list(ids_todo)[:2], 
                         ocr_kwargs=dict(add_img_desc=False), verbose=False, model='claude-sonnet-4-5')
#result

iomeval.pipeline - INFO - 520b24ab... awaiting_curation


iomeval.pipeline - INFO - a1df1fcf... awaiting_curation


iomeval.pipeline - INFO - 05e79e26... awaiting_curation


iomeval.pipeline - INFO - 7ac1b169... awaiting_curation


iomeval.pipeline - INFO - 231001cd... awaiting_curation


iomeval.pipeline - INFO - b00abee5... awaiting_curation


iomeval.pipeline - INFO - 58197a22... awaiting_curation


iomeval.pipeline - INFO - 855ca100... awaiting_curation


iomeval.pipeline - INFO - 0e812bc6... awaiting_curation


iomeval.pipeline - INFO - 1d57797a... awaiting_curation


<iomeval.pipeline.BatchResult>

I've run the ocr and the fixing of the markdown headings multiple times now and would like to further improve this prompt (in particular also that now I can get access to opus with reasoning, ...). If you need further context on how this prompt is used, you can check: https://fr.anckalbi.net/mistocr/refine.html#fix_hdgs

Now here are some situations I've regularly observed why cleaning those markdown:
- I am running it on reports that have a overarching structure (but with a lot of variation) such as:

```
Main title
Executive Summary
TOC (could be before or after the executive summary)
List of figures
List of Abbreviations
Introduction/background
Findings
Conclusions
Recommendations
Annexes
````
The main title is often level 1 while all other level 2 with further subsubsections

- Here are some errors I often observe:
    - The subsection of the TOC are considered as the section/subsection of the report itself and not items of the TOC per se (so the LLM should be very cautious not mixing up a TOC item with a heading of the report per se
    - Some reports might have proper numbering in their headings while others no. Or with a mix of letter based, number based or even without numbering at all. Sometimes table title, or bold text to highlight a sub-sub topic get identified as a heading by the model, ...

- so in summary it's pretty wild. Do you think we could further improve the prompt? Also leveraging the latest opus, reasoning, ...?


For sure "high" should help but shouldn't we start with "low" first?

In [ ]:
!ls ../../iomeval

__init__.py  _modidx.py  curator.py	 extract.py  mapper.py	  readers.py
__pycache__  core.py	 downloaders.py  files	     pipeline.py  themes.py


Look at "pipeline.py" we can now pass dedicated args (kwargs) through fix_kwargs, desc_kwargs.
I'd like to pass a specific model through fix_kwargs and also a specific prompt. Check also this mistocr module: https://github.com/franckalbinet/mistocr/blob/main/nbs/01_refine.ipynb

Yes look at it. But indeed, I remember we can also pass "add_img_desc" directly

Don't you think we could also pass the "reasoning_effort".

## Legacy

In [ ]:
reports = [
    {
        'id': 0,
        'title': 'Better Migration Management Programme Phase III',
        'url':	'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Report%20without%20Annex_Evaluation%20IOM%20BMM%20Phase%20III_26-03-2025.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. BACKGROUND AND CONTEXT',
            '# 4. CONCLUSIONS',
            '# 5. RECOMMENDATIONS'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 1,
        'title': 'Evaluation of IOM Accountability to Affected Populations', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/AAP%20Evaluation%20Report_final_.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## Evaluation Context',
            '## Evaluation Purpose, Objectives and Scope',
            '# 3. CONCLUSIONS',
            '# 4. RECOMMENDATIONS'],
        'evaluator': 'Davide'
    },
    {
        'id': 2,
        'title': 'Final Evaluation of the EU-IOM Joint Initiative for migrant protection and reintegration in the Horn of Africa', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Final%20Evaluation%20Report%20Final_Olta%20NDOJA.pdf',
        'hoi': [
            '## Executive summary',
            '## Methodology',
            '## Findings',
            '## Conclusions and recommendations',
            '## 9. Conclusions',
            '## 10. Recommendations'],
        'evaluator': 'Davide'
    },
    {
        'id': 3,
        'title': 'Evaluation of Mental Health and Psychosocial Support in IOM', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Final%20Report%20Evaluation%20of%20MHPSS%20in%20IOM%208%202024.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## Objective of the Evaluation',
            '### ***Overall Objective of the Evaluation:***',
            '## Scope of the Evaluation',
            '# 4. CONCLUSIONS',
            '# 5. RECOMMENDATIONS'],
        'evaluator': 'Davide'
    },
    {
        'id': 4,
        'title': "Impact evaluation of the UN Secretary General's Peacebuilding Fund-supported East Darfur Assalaya-Sheiria-Yassin Triangle of Peace and Coexistence project", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IE139-PBF-PeaceFIELD-Darfur-report.pdf',
        'hoi': [
            '## Summary',
            '# 1. Introduction',
            '# 7. Discussion',
            '# 8. Conclusion'],
        'evaluator': 'Davide'
    },
    {
        'id': 5,
        'title': "EVALUATION OF IOM'S MIGRATION DATA STRATEGY", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf',
        'hoi': [],
        'evaluator': 'Davide'
    },
    {
        'id': 6,
        'title': 'MALDIVES: STRENGTHENING GOVERNMENT CAPACITY IN MIGRATION HEALTH POLICY DEVELOPMENT', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/EvaluationDocuments_Evaluation_Report_MA.0457_Maldives.pdf',
        'hoi': [
            '## SUMMARY OF THE EVALUATION ASSIGNMENT',
            '## 1. PURPOSE, SCOPE AND OBJECTIVES OF THE SERVICE',
            '## 2. APPROACH AND METHODOLOGY'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 7,
        'title': 'EX-POST EVALUATION OF THE PROJECT: SRI LANKA: UNDERSTANDING MIGRATION, ENVIRONMENTAL DEGRADATION AND CLIMATE CHANGE', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/EvaluationDocuments_NC.0046%20Evaluation%20Report_final.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## 1.1. CONTEXT AND PURPOSE OF THE EVALUATION',
            '## 1.2 EVALUATION BACKGROUND, SCOPE AND PURPOSE',
            '# 3. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 8,
        'title': 'Evaluation of the project “Regional Migration Policy and Knowledge Management Hub in South America”', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Evaluaci%C3%B3n%20PO.0175%20-%20Informe%20final.pdf',
        'hoi': [
            '# 1. EXECUTIVE SUMMARY',
            '## PROJECT BACKGROUND',
            '# 2. CONCLUSIONS',
            '# 3. RECOMMENDATIONS',
            '# 4. LESSONS LEARNED',
            '# 5. GOOD PRACTICES',
            '# 7. CONTEXT AND PURPOSE OF THE EVALUATION'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 9,
        'title': 'FINAL EVALUATION OF THE EU-IOM KNOWLEDGE MANAGEMENT HUB (KMH)', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/KMH%20Final%20Evaluation%20report_final_Olta%20NDOJA.pdf',
        'hoi': [
            '# Executive Summary',
            '# 1. Introduction',
            '# 2. The Evaluation',
            '# 4. Key lessons learned',
            '# 5. Conclusions',
            '# 6. Recommendations'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 10,
        'title': 'Final Internal Evaluation for Enhancing Knowledge on Remittances and Diaspora Engagement in South Sudan', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CD.0011%20evaluation%20report_final_Laura%20SMITH.pdf',
        'hoi': [
            '## Executive Summary',
            '## Evaluation Context and Purpose',
            '## Evaluation Methodology',
            '## Findings and Conclusions',
            '### Lessons learned:',
            '### Good practice:',
            '## Recommendations',
            '## Evaluation context and purpose',
            '## Evaluation Context',
            '### Evaluation Purpose',
            '## Evaluation Scope',
            '## Evaluation Criteria',
            '## Evaluation Framework and Methodology',
            '## Conclusions',
            '## Recommendations'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 11,
        'title': 'Midterm Evaluation of IOM East and Horn of Africa Regional Strategy 2020 – 2024', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Eval%20EHoA%20Regional%20Strategy%20Sept%202023%20final.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. CONTEXT AND PURPOSE OF THE EVALUATION',
            '# 3. LESSONS LEARNED AND GOOD PRACTICES',
            '# 4. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 12,
        'title': "THEMATIC EVALUATION OF IOM'S LABOUR MIGRATION AND MOBILITY STRATEGY AND INITIATIVES", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20LMI%20-%20final%20report%20April%2023.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. INTRODUCTION',
            '# 5. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Andres'
    }
]

In [ ]:
#def get_report_urls():
#    data = httpx.get('https://evaluation.iom.int/json/api/evaluation').json()
#    return {f"https://evaluation.iom.int{e['download_url']}" for e in data['Evaluations']}

In [ ]:
#report_urls = get_report_urls()

In [ ]:
#len(report_urls)

In [ ]:
#def get_eval_report_url(r, report_urls):
#    for d in r.docs:
#        if d['url'] in report_urls:
#            return d['url']
#    return None

In [ ]:
len([e for e in evals if e.meta['Year'] == 2024 ])

44

In [ ]:
reps_2025 = [e for e in evals if e.meta['Year'] >= 2025]
reps_2025[0]


### EVALUATION OF IOM’S MIGRATION DATA STRATEGY
**Year:** 2025 | **Organization:** IOM | **Countries:** Worldwide

**Documents:** 6 available  
**ID:** `9992310969aa2f428bc8aba29f865cf3`


In [ ]:
from iomeval.pipeline import run_pipeline

In [ ]:
reports = [
    {
        'id': 0,
        'title': 'Better Migration Management Programme Phase III',
        'url':	'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Report%20without%20Annex_Evaluation%20IOM%20BMM%20Phase%20III_26-03-2025.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. BACKGROUND AND CONTEXT',
            '# 4. CONCLUSIONS',
            '# 5. RECOMMENDATIONS'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 1,
        'title': 'Evaluation of IOM Accountability to Affected Populations', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/AAP%20Evaluation%20Report_final_.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## Evaluation Context',
            '## Evaluation Purpose, Objectives and Scope',
            '# 3. CONCLUSIONS',
            '# 4. RECOMMENDATIONS'],
        'evaluator': 'Davide'
    },
    {
        'id': 2,
        'title': 'Final Evaluation of the EU-IOM Joint Initiative for migrant protection and reintegration in the Horn of Africa', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Final%20Evaluation%20Report%20Final_Olta%20NDOJA.pdf',
        'hoi': [
            '## Executive summary',
            '## Methodology',
            '## Findings',
            '## Conclusions and recommendations',
            '## 9. Conclusions',
            '## 10. Recommendations'],
        'evaluator': 'Davide'
    },
    {
        'id': 3,
        'title': 'Evaluation of Mental Health and Psychosocial Support in IOM', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Final%20Report%20Evaluation%20of%20MHPSS%20in%20IOM%208%202024.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## Objective of the Evaluation',
            '### ***Overall Objective of the Evaluation:***',
            '## Scope of the Evaluation',
            '# 4. CONCLUSIONS',
            '# 5. RECOMMENDATIONS'],
        'evaluator': 'Davide'
    },
    {
        'id': 4,
        'title': "Impact evaluation of the UN Secretary General's Peacebuilding Fund-supported East Darfur Assalaya-Sheiria-Yassin Triangle of Peace and Coexistence project", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IE139-PBF-PeaceFIELD-Darfur-report.pdf',
        'hoi': [
            '## Summary',
            '# 1. Introduction',
            '# 7. Discussion',
            '# 8. Conclusion'],
        'evaluator': 'Davide'
    },
    {
        'id': 5,
        'title': "EVALUATION OF IOM'S MIGRATION DATA STRATEGY", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf',
        'hoi': [],
        'evaluator': 'Davide'
    },
    {
        'id': 6,
        'title': 'MALDIVES: STRENGTHENING GOVERNMENT CAPACITY IN MIGRATION HEALTH POLICY DEVELOPMENT', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/EvaluationDocuments_Evaluation_Report_MA.0457_Maldives.pdf',
        'hoi': [
            '## SUMMARY OF THE EVALUATION ASSIGNMENT',
            '## 1. PURPOSE, SCOPE AND OBJECTIVES OF THE SERVICE',
            '## 2. APPROACH AND METHODOLOGY'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 7,
        'title': 'EX-POST EVALUATION OF THE PROJECT: SRI LANKA: UNDERSTANDING MIGRATION, ENVIRONMENTAL DEGRADATION AND CLIMATE CHANGE', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/EvaluationDocuments_NC.0046%20Evaluation%20Report_final.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '## 1.1. CONTEXT AND PURPOSE OF THE EVALUATION',
            '## 1.2 EVALUATION BACKGROUND, SCOPE AND PURPOSE',
            '# 3. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Davide'
    },
    {
        'id': 8,
        'title': 'Evaluation of the project “Regional Migration Policy and Knowledge Management Hub in South America”', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Evaluaci%C3%B3n%20PO.0175%20-%20Informe%20final.pdf',
        'hoi': [
            '# 1. EXECUTIVE SUMMARY',
            '## PROJECT BACKGROUND',
            '# 2. CONCLUSIONS',
            '# 3. RECOMMENDATIONS',
            '# 4. LESSONS LEARNED',
            '# 5. GOOD PRACTICES',
            '# 7. CONTEXT AND PURPOSE OF THE EVALUATION'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 9,
        'title': 'FINAL EVALUATION OF THE EU-IOM KNOWLEDGE MANAGEMENT HUB (KMH)', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/KMH%20Final%20Evaluation%20report_final_Olta%20NDOJA.pdf',
        'hoi': [
            '# Executive Summary',
            '# 1. Introduction',
            '# 2. The Evaluation',
            '# 4. Key lessons learned',
            '# 5. Conclusions',
            '# 6. Recommendations'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 10,
        'title': 'Final Internal Evaluation for Enhancing Knowledge on Remittances and Diaspora Engagement in South Sudan', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CD.0011%20evaluation%20report_final_Laura%20SMITH.pdf',
        'hoi': [
            '## Executive Summary',
            '## Evaluation Context and Purpose',
            '## Evaluation Methodology',
            '## Findings and Conclusions',
            '### Lessons learned:',
            '### Good practice:',
            '## Recommendations',
            '## Evaluation context and purpose',
            '## Evaluation Context',
            '### Evaluation Purpose',
            '## Evaluation Scope',
            '## Evaluation Criteria',
            '## Evaluation Framework and Methodology',
            '## Conclusions',
            '## Recommendations'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 11,
        'title': 'Midterm Evaluation of IOM East and Horn of Africa Regional Strategy 2020 – 2024', 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Eval%20EHoA%20Regional%20Strategy%20Sept%202023%20final.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. CONTEXT AND PURPOSE OF THE EVALUATION',
            '# 3. LESSONS LEARNED AND GOOD PRACTICES',
            '# 4. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Andres'
    },
    {
        'id': 12,
        'title': "THEMATIC EVALUATION OF IOM'S LABOUR MIGRATION AND MOBILITY STRATEGY AND INITIATIVES", 
        'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20LMI%20-%20final%20report%20April%2023.pdf',
        'hoi': [
            '# EXECUTIVE SUMMARY',
            '# 1. INTRODUCTION',
            '# 5. CONCLUSIONS AND RECOMMENDATIONS'
            ],
        'evaluator': 'Andres'
    }
]

In [ ]:
reps_completed = [find_eval(evals, o['url'], by='url').id for o in reports]
reps_completed

['4341695461234eee3deb51ac68871109',
 '6c3c2cf3fa479112967612b0baddab72',
 '49d2fba781b6a7c0d94577479636ee6f',
 '22cac1c000836253adc445993e101560',
 'e8d9427f4a9094f68e17307faa567d8b',
 '9992310969aa2f428bc8aba29f865cf3',
 'a493c447cff3e03ca94024335d5c5c7a',
 '06fd0949dc874f388686461c19d2669c',
 '4c04ec0d3e651f7616190c6fc2dcfc18',
 '891448b7a4bf12763ea0579dd67e289d',
 '75bad4665ed458cce78740e3d86f72d0',
 'f409b3f9a0671ec69a8591c425a45c1d',
 '760568a15e144bcfb88dbbd50247dc51']

In [ ]:
[o.id for o in reps_2025 if o.id not in reps_completed]

['2b297bc904ed1e542fbe67f62a8ea7cf',
 '227dbf5b243f3973312b7de3be53105d',
 '7a6f58a2fa72c3290ceae54f0924ea18',
 'f9ff95796dd18443fcc94fd6b795acf6',
 '8faad6edb0414cfc8f35415a93726e36',
 '3d2c35aec5bde93fc1db65884af3daa8',
 '9591f983c436297929b9e240a3c28df6',
 '316c47c9dc19c820f3217434d4b93b28',
 'e21305a89aeb76fafa8f4270392d463a',
 '1700c8dbadc3d87d6911a8ccc6d18b63',
 '0d4309b702e00faac631be8d609934ce',
 '7c89660d359efcf0f8bba26f79ae8071',
 'cf4fcf44e2eff305339aeeb5d2290f7f',
 '48fb10f3895036743cb8811daac46bcd']

In [ ]:
r = find_eval(evals, '2b297bc904ed1e542fbe67f62a8ea7cf', by='id')
r


### Meta-evaluation of IOM evaluations 2020–2024
**Year:** 2025 | **Organization:** IOM | **Countries:** Worldwide

**Documents:** 3 available  
**ID:** `2b297bc904ed1e542fbe67f62a8ea7cf`


In [ ]:
lut_title_url[r.meta['Title']]

'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Meta-evaluation%202020-2024%20report%20-%20clean.pdf'

In [ ]:
pdf_links = {
    '2b297bc904ed1e542fbe67f62a8ea7cf': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Meta-evaluation%202020-2024%20report%20-%20clean.pdf',
    '227dbf5b243f3973312b7de3be53105d': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PBF%20SPSC%20IOM%20UNDP_Final%20Evaluation%20Report.pdf',
    '7a6f58a2fa72c3290ceae54f0924ea18':'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/WBRR%20Phase%20II%20Final%20Evaluation%20Report%20%2827%20June%202025%29%20%281%29.pdf',
    'f9ff95796dd18443fcc94fd6b795acf6': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Evaluation%20Report_Theogene%20NSHIMIYIMAN.pdf',
    '8faad6edb0414cfc8f35415a93726e36':'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CS1137-CS1138_Final%20Report_French_Javier%20SANCHEZ%20MARTI.pdf',
    '3d2c35aec5bde93fc1db65884af3daa8': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CPCD-%20KfW%20Phase%20III%20Final%20Evaluation%20Report.._Menna%20Allah%20ALAMEN.pdf',
    '9591f983c436297929b9e240a3c28df6': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/WBCAP_WBIBM%20Evaluation%20Report%20Clean.pdf',
    '316c47c9dc19c820f3217434d4b93b28': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CBI%20Evaluation_final%20evaluation%20report.pdf',
    'e21305a89aeb76fafa8f4270392d463a': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IAHE%20of%20the%20Drought%20Response%20in%20Somalia.pdf',
    '1700c8dbadc3d87d6911a8ccc6d18b63': None,
    '0d4309b702e00faac631be8d609934ce': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOMFund%20evaluation_final%20report.pdf',
    '7c89660d359efcf0f8bba26f79ae8071': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Senegal%20MA.0464_Final%20Evaluation%20Report%2012.09_Soutongnooma%20Yves%20Ab.pdf',
    'cf4fcf44e2eff305339aeeb5d2290f7f': None, # too big to be further processed

}

In [ ]:
for k, v in pdf_links.items():
    if v:
        r = find_eval(evals, k, by='id')
        found = get_eval_report_url(r, report_urls)
        print(r.id, found == v, r.meta['Title'][:50])

2b297bc904ed1e542fbe67f62a8ea7cf True Meta-evaluation of IOM evaluations 2020–2024
227dbf5b243f3973312b7de3be53105d True Final evaluation of the project: Sustaining peace 
7a6f58a2fa72c3290ceae54f0924ea18 True WESTERN BALKANS ASSISTED VOLUNTARY RETURN AND REIN
f9ff95796dd18443fcc94fd6b795acf6 True Final Evaluation of the Project "Draft the Future:
8faad6edb0414cfc8f35415a93726e36 False Renforcement du dialogue binational entre Haïti et
3d2c35aec5bde93fc1db65884af3daa8 True IRAQ’S ECONOMIC RECOVERY THROUGH EMPLOYMENT CREATI
9591f983c436297929b9e240a3c28df6 True Building Facility (WBIBM) & Western Balkans Readmi
316c47c9dc19c820f3217434d4b93b28 True EVALUATION OF IOM CASH-BASED INTERVENTIONS (CBI)
e21305a89aeb76fafa8f4270392d463a True INTER-AGENCY HUMANITARIAN EVALUATION OF THE RESPON
0d4309b702e00faac631be8d609934ce False Fifth Evaluation of the IOM Development Fund
7c89660d359efcf0f8bba26f79ae8071 True Evaluation ex-post interne du projet « Renforcemen
cf4fcf44e2eff305339aeeb5d2290f7f 

In [ ]:
pdf_links[r.id]

'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/CoFunding%20Mechanism%20Internal%20Evaluation%20Repor_Elise%20Caroline%20Anais.pdf'

In [ ]:
report = await run_pipeline(evals, url=pdf_links[r.id], base_path='../../data', ocr_kwargs={'add_img_desc': False})

iomeval.pipeline - INFO - Running OCR...


mistocr.pipeline - INFO - Step 1/2: Running OCR on ../../data/pdf/cf4fcf44e2eff305339aeeb5d2290f7f/CoFunding%20Mechanism%20Internal%20Evaluation%20Repor_Elise%20Caroline%20Anais.pdf...


SDKError: API error occurred: Status 413 Content-Type "application/json; charset=utf-8". Body: {
  "message":"Request size limit exceeded",
  "request_id":"c500ca3f6c165c72bf1c555442768b60"
}

## Pipeline dev.

In [ ]:
evals[0], len(evals)

(Evaluation(id='9992310969aa2f428bc8aba29f865cf3', docs=[{'subtype': 'Evaluation report', 'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Annex%20VI%20Case%20Study%20-%20RDH%20East%2C%20Horn%20and%20Southern%20Africa.pdf', 'desc': 'Evaluation Report'}, {'subtype': 'Evaluation brief', 'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Annex%20VII%20Case%20Study%20-%20RDH%20Asia-Pacific.pdf', 'desc': 'Evaluation Brief'}, {'subtype': 'Annexes', 'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Annex%20VIII%20-%20Inception%20Report.pdf', 'desc': 'Annex VI Case Study - RDH East'}, {'subtype': 'Annexes', 'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/Evaluation%20Brief.pdf', 'desc': 'Horn and Southern Africa'}, {'subtype': 'Annexes', 'url': 'https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf', 'desc'

In [ ]:
!ls ..

00_core.ipynb	      04_themes.ipynb	 _index_dup1.ipynb  index.ipynb
01_readers.ipynb      05_mapper.ipynb	 _quarto.yml	    nbdev.yml
02_downloaders.ipynb  06_curator.ipynb	 analysis	    sidebar.yml
03_extract.ipynb      07_pipeline.ipynb  files		    styles.css


My goal is to run a LLM-based pipeline as developped and described in "07_pipeline.ipynb". I have to do it for every report of "evals" (740 reports in total). As you can see it's a multi-step pipeline: from pdf download, ocr, fixing headings, manual curation of generated markdown version of the pdf report and manual selection of subsections of interest (through "06_curator.ipynb" fasthtml app), then mapping against various evaluation framework. 

For each report being processed, we create an associated json file such as: 
```
{
  "id": "227dbf5b243f3973312b7de3be53105d",
  "report_url": "https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PBF%20SPSC%20IOM%20UNDP_Final%20Evaluation%20Report.pdf",
  "meta": {
    "Title": "Final evaluation of the project: Sustaining peace and social cohesion in Bosnia and Herzegovina through enhanced inter-municipal and inter-entity cooperation on local services (SPSC)",
    "Year": 2025,
    "Author": "CREDI",
    "Best Practicesor Lessons Learnt": "Yes",
    "Date of Publication": "2025-06-30",
    "Donor": "UN Peace Fund",
    "Evaluation Brief": "Yes",
    "Evaluation Commissioner": "Donor",
    "Evaluation Coverage": "Country",
    "Evaluation Period From Date": NaN,
    "Evaluation Period To Date": NaN,
    "Executive Summary": "Yes",
    "External Version of the Report": "No",
    "Languages": "English",
    "Migration Thematic Areas": "Transition Recovery - peacebuilding and peace preservation, Transition Recovery - preventing violent extremism",
    "Name of Project(s) Being Evaluated": NaN,
    "Number of Pages Excluding annexes": NaN,
    "Other Documents Included": NaN,
    "Project Code": "PB0087",
    "Countries Covered": [
      "Bosnia and Herzegovina"
    ],
    "Regions Covered": "RO for Europe and Central Asia",
    "Relevant Crosscutting Themes": "Gender",
    "Report Published": "Yes",
    "Terms of Reference": "Yes",
    "Type of Evaluation Scope": "Programme/Project",
    "Type of Evaluation Timing": "Final (at the end of the project/programme)",
    "Type of Evaluator": "External",
    "Level of Evaluation": "Decentralized",
    "Document Subtype": "Evaluation report, Evaluation brief, Management response",
    "File URL": "https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PB0087_Evaluation_Management%20response.pdf,   https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PBF%20SPSC%20IOM%20UNDP%20Project_Evaluation%20brief.pdf,   https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PBF%20SPSC%20IOM%20UNDP_Final%20Evaluation%20Report.pdf",
    "File description": "Evaluation Report , Evaluation Brief, Management Response",
    "Management response": "Yes",
    "Date added": "Thu, 08/28/2025 - 22:59",
    "report_url": "https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PBF%20SPSC%20IOM%20UNDP_Final%20Evaluation%20Report.pdf"
  },
  "docs": [
    {
      "subtype": "Evaluation report",
      "url": "https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PB0087_Evaluation_Management%20response.pdf",
      "desc": "Evaluation Report"
    },
    {
      "subtype": "Evaluation brief",
      "url": "https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PBF%20SPSC%20IOM%20UNDP%20Project_Evaluation%20brief.pdf",
      "desc": "Evaluation Brief"
    },
    {
      "subtype": "Management response",
      "url": "https://evaluation.iom.int/sites/g/files/tmzbdl151/files/docs/resources/PBF%20SPSC%20IOM%20UNDP_Final%20Evaluation%20Report.pdf",
      "desc": "Management Response"
    }
  ],
  "curation_status": "sections_selected",
  "selected_headings": [
    "## EXECUTIVE SUMMARY ... page 4",
    "## INTRODUCTION ... page 9",
    "## EVALUATION PURPOSE AND OBJECTIVES ... page 12",
    "## CONCLUSIONS ... page 67",
    "## LESSONS LEARNED AND RECOMMENDATIONS ... page 69"
  ],
  "mappings": {},
  "timestamp": "2026-02-06T13:50:26.995375"
}
```

When running on a subset of reports (e.g all 2025 or 2024, ...) some might suceed, other fail (pdf too large > 50MB, LLM api error, ...). When I run it multiple time I might need to force re-running some of the steps (mapping for instance or ocr) but should always be intentional. Also we'd like to keep potentially (for the mappings itself) previous versions, ... 

Concretelt, if for instance I run it all on 2024 (about 45 reports), first pdf downloads, ocr, ... will be applied then I'll have to curate the produced markdown, select sections, ... then I can resume the pipeline with the mapping, ... But the next time I want to map, I won't have to go through the curation again, just the mapping. To do so I can force the mapping (I have flags for that in my run_pipeline functions, ...). At the same time I want to keep previous versions, ... So that I can monitor also how mappings differ over run...

So there is this "complex" worklow/orchestration I want to further think about so that eventually it's clear and clean how to iterate, regenerate mappings for whatever reasons (updated prompts, new reports, generating progressively mapping for older and older reports, monitoring what failed (why?) and not), ...

- Let me clarify one point, currently in the json per report named using the id of the report (e.g `227dbf5b243f3973312b7de3be53105d.json`), we get those two attributes "curation_status", and the "selected_headings" to use to extract a subset of the report. For instance:

```
...
"curation_status": "sections_selected",
  "selected_headings": [
    "## EXECUTIVE SUMMARY ... page 4",
    "## INTRODUCTION ... page 9",
    "## EVALUATION PURPOSE AND OBJECTIVES ... page 12",
    "## CONCLUSIONS ... page 67",
    "## LESSONS LEARNED AND RECOMMENDATIONS ... page 69"
  ],
...
```

So in the "curator" app we know which one to curate or what needs to be curated. Whe now the status ('pending', 'headings_reviewed', 'sections_selected'). 'pending' means we went through the ocr already, 'headings_reviewed' that we cleaned the markdown headings while 'selection_selected' says that we are ready for mapping

- for the versionning question, we could simply archive a copy of those json in a timestamped directory `/20260105`?
- for the batch, yes we could have batch function like `batch_run` to run particular subset of the evaluation reports while tracking which one succeeded or not
- let's further discuss "idempotency". It means same effect if run multiple times? Let's further discuss the use case
- I think I already answered the question on "manual curation workflow"


Don't you think it's ok to retry manually? This will also help better understand the failure modes and getting acquainted with the input data, ...?

- the json report are handled by the "run_pipeline" function right? I see this "run_pipeline" as an "inner loop" per report while seeing the orchestration/batch as an outer loop (not sure that's the right way seeing it though ...). I am leaning toward option 3 as something that see what failed at a "macro" level...

Yes makes sense

It looks fine

I think we should start with review/refactoring `run_pipeline` to ensure it plays well. Also note that I might want to delete the pdf report downloaded once ocr applied successfully (to save local space and they are not needed later on as I can always get access them online)

Let's work on "value clarity" first indeed.

A dataclass looks fine though fastcore might have its own way to implement data class right?

Ok whatever cleaner and simpler is fine to me

Yes show me

Looks great! A few remarks for the helper though
- remember that I wanted to delete the pdf once ocr succeeded so we might check the md_path (if none, we'll have to re-download it)

yes

Ok added!

yes option b better 

Can't I delete a file using fastcore?

Not that usually the pdf file is downloaded in a folder under "pdf/id-report/" for instance. I think that the `mistocr` is taking care of that: https://raw.githubusercontent.com/franckalbinet/mistocr/refs/heads/main/mistocr/core.py

Where to add this logic? I'd lean toward putting it in the `ocr` function in the "07_pipeline.ipynb" but not sure

Right! Can you show me option 2 then? (just the parts to update)

Ok, in such a case we delete the pdf only not the parent folder

Ok let's further examine it. Currently we have:
- pdf_url: which is THE online link of the pdf report of interesed and associated with an evaluation
- pdf_path: which is the returned by `self.pdf_path = download_eval(self.ev, dst=dst)` in the `download` method of the pipeline
- in the jsons themselves (per report), we store the online link of the pdf as `report_url` 

Note also that in the `_repr_markdown_` we use pdf_path to show the "downloaded" status (not sure useful then)


Let's think about it. What counts to us it the conversion process to markdown. Furthermore having an md implies that we have downloaded at one point. And last, if we need to go through the ocr again, we can just download it again. Pdf download is just a mean not an end

Agree! Actually, we are not persisting "pdf_path" to the json currently (check the example shown earlier). Can you show the updated "_repr_markdown_" and whatever needs to be updated?

Great, what's next?

Wait I've not yet implemented the delete_pdf step in the code base. So what should I do finally?

Wait, let's be cautious here and further assess how this `pdf_path` is handled. In particular look at:
the `load_report` function. While runnin the pipeline for a single report, we get the pdf_path from the `download` method, go ahead with the download, `save` the report (without explicitely store the pdf_path, ... The next time we load a Report, we infer this path from the base_path in the `load_report` function:
```
...
pdf_dir = Path(base_path)/'pdf'/id
    md_dir = Path(base_path)/'md'/id
    report.pdf_path = pdf_dir if pdf_dir.exists() else None
    report.md_path = md_dir if md_dir.exists() else None
...```

Don't you think anything to improve, to further pay attention to?

On my side, would lean toward option A. What do you think?

Ok, updated! What's next?

Ok let's implement `batch_run`

- succeeded: obviously useful
- failed: definitely important to know why - pdf too big, ocr or mapping error, ...
- skipped: what could be the use case?

No those information for "skipped" might be useful indeed.
Let's try to validate the approach with a few examples:
- let's run batch for all 2023 reports, none were processed before, so will get succeeded or failed. But only ocr will have succeeded so far, curation still needed. How/should we differentiate what succeeded also? We can always infer that though through the json of a report...
- let's re-run the same batch, some have been successfully processed (partially or not, ...). How do we proceed here, we re-run and resume pipeline where it stopped last time (unless we pass force options, ...)

Yes makes sense to mirror PipelineResult and use "skipped" only in the case mentioned. For instance what would happen if I ran the 2024 batch but some were already processed throughout before?

Ok let's tackle it step by step

Ok, then?

good, then?

Not sure. I think we can infer the state from the json. After all, those json are a way to save the state right?

Yes right if we get mappings for "enbs", "ccps", "gcms" and "outs", then we are done!

Yes

Fine! Just note that currently we run_pipeline using the `url` way. Indeed the current evals docs are messed up so I have to use the "get_eval_report_url" method to look it up. And yes sometimes it won't find the url, so I'll have to capture this error mode also

Yes tentativaly until we get a fixed evaluation repository

Yes

Yes let's tackle persistence now

1. Yes base_path makes sense
2. I do not mind. what would be the best practice?

Fair enough. 

I think it's fine. 

Can you show me the code for the batch results along with its method save? 

I think I have implemented all in the "07_pipeline.ipynb" module (around the end of it). Can you double check? Once done, we have to test it at small scale  

## Test batch

The reports in $`reports` are reports already mapped but with different prompts, ocr approach, json dump structure so I'll have to run the pipeline again anyway. They are some reports from 2025, 2024 that I excluded in the current base_path 

Their previous json dump where store somewhere else so no need to force anything. Furthermore, we can grap them by id directly and test our tentative way to retrieve the pdf url also.

In [ ]:

find_eval??


```python
def find_eval(
    evals:list, # List of evaluations
    query:str, # Title or URL of evaluation
    by:str='title' # 'title', 'url' or 'id'
    ): 
    "Find evaluation by title, URL or id"
    if by == 'title': return first([o for o in evals if o.meta['Title'] == query])
    if by == 'url': return first([o for o in evals if in_docs(o, query)])
    if by == 'id': return first([o for o in evals if o.id == query])
```

**File:** `~/iomeval/iomeval/readers.py`

Look we can find evals by id also

In [ ]:
test_ids = reps_completed[:3]
test_ids

['4341695461234eee3deb51ac68871109',
 '6c3c2cf3fa479112967612b0baddab72',
 '49d2fba781b6a7c0d94577479636ee6f']

In [ ]:
result = await batch_run(evals, report_urls, base_path='../../data', ids=test_ids)
result

Can we reimport above those two cells (test) everything required to be able to run "batch_run"?

Wouldn't be better to test it directly from the "07_pipeline.ipynb" module?

Show me the whole code reuired to be able to test it

I think we'll need to add get_eval_report_url and its dependencies in the module also

## Legacy

In [ ]:
from mistralai import Mistral
import os

c = Mistral(api_key=os.getenv("MISTRAL_API_KEY"))

In [ ]:
c.batch.jobs.cancel(job_id="2fa17c1d-959a-46ae-b0b2-7d966eb2c56a")

BatchJobOut(id='2fa17c1d-959a-46ae-b0b2-7d966eb2c56a', input_files=['3d7c400a-055a-4c9f-ad7a-6b156766f924'], endpoint='/v1/ocr', errors=[], status='CANCELLATION_REQUESTED', created_at=1770371001, total_requests=1, completed_requests=0, succeeded_requests=0, failed_requests=0, object='batch', metadata=None, model='mistral-ocr-latest', agent_id=None, output_file=None, error_file=None, started_at=1770371002, completed_at=None)

In [ ]:
job = c.batch.jobs.get(job_id="e7d64ad4-8394-4c9a-a3f4-d36a1681afcd")
print(job.status, job)

CANCELLED id='e7d64ad4-8394-4c9a-a3f4-d36a1681afcd' input_files=['96c40ca9-fbde-40e9-80f5-d6a6c0207cb9'] endpoint='/v1/ocr' errors=[] status='CANCELLED' created_at=1770310312 total_requests=1 completed_requests=1 succeeded_requests=1 failed_requests=0 object='batch' metadata=None model='mistral-ocr-latest' agent_id=None output_file='4e0a858e-6a7d-42b5-aa6f-fdcf56bde6d2' error_file=None started_at=1770310313 completed_at=1770313287


Ok ran those commands. What do you think?

But two days ago, it worked in 20 s

Can we try the synchronous mistral OCR?

In [ ]:
jobs = c.batch.jobs.list()
for j in jobs.data[:10]: print(j.id, j.status, j.created_at)

Do you think it's reasonable that during busy hours, running might take minutes for a single pdf file (200 pages)

In [ ]:
import os
from urllib.parse import unquote
Path(f"../../data/pdf/{r.id}").ls()

[Path('../../data/pdf/cf4fcf44e2eff305339aeeb5d2290f7f/CoFunding%20Mechanism%20Internal%20Evaluation%20Repor_Elise%20Caroline%20Anais.pdf')]

What do you think?

In [ ]:
c.batch.jobs.cancel(job_id="944d4d93-271e-4994-8c72-dd1cf54d1182")

BatchJobOut(id='944d4d93-271e-4994-8c72-dd1cf54d1182', input_files=['1659ac3c-db21-4ffa-b11a-ae053621c921'], endpoint='/v1/ocr', errors=[], status='CANCELLATION_REQUESTED', created_at=1770307352, total_requests=1, completed_requests=0, succeeded_requests=0, failed_requests=0, object='batch', metadata=None, model='mistral-ocr-latest', agent_id=None, output_file=None, error_file=None, started_at=1770307352, completed_at=None)

In [ ]:
jobs = c.batch.jobs.list()
for j in jobs.data[:10]: print(j.id, j.status, j.created_at)

9683cafb-a3a4-4346-934b-e6b06865657e CANCELLED 1770307850
944d4d93-271e-4994-8c72-dd1cf54d1182 CANCELLED 1770307352
71837aaf-94fc-4683-970c-f205dd3a4543 SUCCESS 1770200796
9bc2fe43-e703-40a1-b992-ad16d558b8bd SUCCESS 1770200560
a4b4d6c5-5df6-41a4-892e-696ef985e73d SUCCESS 1770195698
f420c186-3b24-4239-a0b9-edfce94ef160 SUCCESS 1770156065
a4b6afb4-6a99-4314-aa13-142838abfa40 SUCCESS 1770153179
0a7bf80a-2b53-48d2-a209-e96ffefb6711 SUCCESS 1770152568
d3c90db1-a159-4d69-a91c-a710a9addbd9 SUCCESS 1770142956
44095102-4856-4192-8e4e-c92f2d4c0b47 SUCCESS 1770142000


Ok, here the new status! Should I wait cancellation?

In [ ]:
!ls ../../data/md

227dbf5b243f3973312b7de3be53105d  2b297bc904ed1e542fbe67f62a8ea7cf


In [ ]:
"#International Organization for Migration (IOM)"

'#International Organization for Migration (IOM)'

In [ ]:
md_dir = Path('../../data/md')
pgs = read_pgs(md_dir / r.id, join=False)
pgs[28]

"#\n\nIAAHE\n\n#### Funding ... page 29\n\n62. Information about the humanitarian response's funding situation in Somalia is inconsistent across different sources, for instance, when comparing UN OCHA's Financial Tracking Service (FTS) data and reports for pooled funds. Moreover, as previous reviews and evaluations have emphasised, these reports do not cover the significant resources provided through remittances and businesses.\n63. Figure 7 shows the trends in funding reported in UN OCHA's FTS. Table 1 seeks to provide an approximate overview of relevant figures relating to the IASC system. According to these figures, the 2022-2023 crisis saw a significant but short-lived increase in funding, roughly doubling what was available in each of the previous six years. Along with the funding growth came a shift in funding distribution across sectors (see Chapter 3.1 for details) and a broadening of the response's donor base.\n\nFigure 7: Trends in Reported Funding\n![img-27.jpeg](img-27.jpeg

I think that the "\nThe UN Migration Agency\nCREDI\nCenter su correjoe evaluacije i direktivna intrativna\n\n" part is still junk

Let's do it case by case

In [ ]:
def replace_in_pages(pg_dir, old_str, new_str=''):
    "Replace `old_str` with `new_str` in all markdown files in `pg_dir`, in place."
    pg_dir = Path(pg_dir)
    for f in sorted(pg_dir.glob('*.md')):
        txt = f.read_text()
        f.write_text(txt.replace(old_str, new_str))

In [ ]:
header_block = "#\n\n"
replace_in_pages(md_dir / r.id, header_block)

In [ ]:
#for e in reps_2025:
#    report = await run_pipeline(reps_2025, id=e.id, base_path='../../_data', ocr_kwargs={'add_img_desc': False})
#    print(f"{report.id}: {report.curation_status}")

In [ ]:
!ls ../../data_tmp/pdf_test


IOM%20MDS%20Evaluation%20Report%20-%20clean_0.pdf


In [ ]:
from mistocr.core import ocr_pdf
mds = ocr_pdf('../../data_tmp/pdf_test', '../../data_tmp/md')
mds

mistocr.core - INFO - Waiting for batch job 8dfd6bc7-1506-4ebe-a4c2-38abca4f26cf (initial status: QUEUED)


mistocr.core - DEBUG - Job 8dfd6bc7-1506-4ebe-a4c2-38abca4f26cf status: QUEUED (elapsed: 0s)


mistocr.core - DEBUG - Job 8dfd6bc7-1506-4ebe-a4c2-38abca4f26cf status: RUNNING (elapsed: 2s)


mistocr.core - DEBUG - Job 8dfd6bc7-1506-4ebe-a4c2-38abca4f26cf status: RUNNING (elapsed: 2s)


mistocr.core - DEBUG - Job 8dfd6bc7-1506-4ebe-a4c2-38abca4f26cf status: RUNNING (elapsed: 2s)


mistocr.core - INFO - Job 8dfd6bc7-1506-4ebe-a4c2-38abca4f26cf completed with status: SUCCESS


[Path('../../data_tmp/md/IOM%20MDS%20Evaluation%20Report%20-%20clean_0')]

In [ ]:
!rm -rf ../../data/pdf

Note that there are still occurences of fixes where we have 
```
## 3. subsection title
## 3.1 subsubsection subtitle
```

where the model doesn't identify the issue and that "## 3.1 .." should actually be "### 3.1 ..". How to capture/fix this in a generic way?

Would prefer improving the prompt, because the deterministic will for sure face edge cases, ...

Yes, worth trying. Where to put it?